tools/setup/emoji/emoji_names.py

```diff
@@ -1,1578 +1,34 @@
-    # timezone and locale issues, and uses a digital representation.
-    # '1f550': {'canonical_name': 'X', 'aliases': ['clock1']},
-    # '1f551': {'canonical_name': 'X', 'aliases': ['clock2']},
-    # '1f552': {'canonical_name': 'X', 'aliases': ['clock3']},
-    # '1f553': {'canonical_name': 'X', 'aliases': ['clock4']},
-    # '1f554': {'canonical_name': 'X', 'aliases': ['clock5']},
-    # '1f555': {'canonical_name': 'X', 'aliases': ['clock6']},
-    # '1f556': {'canonical_name': 'X', 'aliases': ['clock7']},
-    # seems like the best choice for time
-    "1f557": {"canonical_name": "time", "aliases": ["clock"]},
-    # '1f558': {'canonical_name': 'X', 'aliases': ['clock9']},
-    # '1f559': {'canonical_name': 'X', 'aliases': ['clock10']},
-    # '1f55a': {'canonical_name': 'X', 'aliases': ['clock11']},
-    # '1f55b': {'canonical_name': 'X', 'aliases': ['clock12']},
-    # '1f55c': {'canonical_name': 'X', 'aliases': ['clock130']},
-    # '1f55d': {'canonical_name': 'X', 'aliases': ['clock230']},
-    # '1f55e': {'canonical_name': 'X', 'aliases': ['clock330']},
-    # '1f55f': {'canonical_name': 'X', 'aliases': ['clock430']},
-    # '1f560': {'canonical_name': 'X', 'aliases': ['clock530']},
-    # '1f561': {'canonical_name': 'X', 'aliases': ['clock630']},
-    # '1f562': {'canonical_name': 'X', 'aliases': ['clock730']},
-    # '1f563': {'canonical_name': 'X', 'aliases': ['clock830']},
-    # '1f564': {'canonical_name': 'X', 'aliases': ['clock930']},
-    # '1f565': {'canonical_name': 'X', 'aliases': ['clock1030']},
-    # '1f566': {'canonical_name': 'X', 'aliases': ['clock1130']},
-    # '1f567': {'canonical_name': 'X', 'aliases': ['clock1230']},
-    "1f3f3": {"canonical_name": "white_flag", "aliases": ["surrender"]},
-    "1f3f4": {"canonical_name": "black_flag", "aliases": []},
-    "1f3c1": {"canonical_name": "checkered_flag", "aliases": ["race", "go", "start"]},
-    "1f6a9": {"canonical_name": "triangular_flag", "aliases": []},
-    # solidarity from iemoji
-    "1f38c": {"canonical_name": "crossed_flags", "aliases": ["solidarity"]},
-}
+import orjson
+
+TOOLS_DIR = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+ZULIP_PATH = os.path.dirname(TOOLS_DIR)
+EMOJI_DATA_FILE_PATH = os.path.join(
+    ZULIP_PATH, "node_modules", "emoji-datasource-google", "emoji.json"
+)
+
+
+def reformat_name(name: str) -> str:
+    return name.lower().replace(" ", "_").replace("-", "_")
+
+
+# TODO this is copied verbatim from emoji_setup_utils.py. We should use that
+# function instead when refactoring the code cruft for our emoji system.
+def get_emoji_code(emoji_dict: Dict[str, Any]) -> str:
+    emoji_code = emoji_dict.get("non_qualified") or emoji_dict["unified"]
+    return emoji_code.lower()
+
+
+with open(EMOJI_DATA_FILE_PATH, "rb") as emoji_data_file:
+    emoji_data = orjson.loads(emoji_data_file.read())
+
+EMOJI_NAME_MAPS: Dict[str, Dict[str, Any]] = {}
+for emoji_dict in emoji_data:
```

Comment:
Another effect of this PR is that we'll add all the emoji that exist in this data source.

There are some emoji we've explicitly left out in the existing list, with reasons for them -- search for `'X'`. For some of those, it may be best to relax and let them in along with a policy of using upstream names, but again that should be an explicit decision and not a side effect of anything.

This may also pull in a large number of emoji that are composed as sequences. For almost any emoji that shows a person, there are (1 + 2) * (1 + 5) = 18 different sequences specifying, or not specifying, gender and skin tone. For emoji like 👪 U+1F46A FAMILY with several people, there are even more sequences. I don't think it would be a good experience in our current typeahead UI for each potentially-relevant emoji to show up in 18+ different versions -- it'd mean you'd effectively see just one possible match, in a few of its possible versions. We want to support these (#992), but I think it will take some design work on the UI to do a good job of them. In any case, again: adding them should be an explicit choice, where we look concretely at the results it produces in the UX.

Reply:
> There are some emoji we've explicitly left out in the existing list, with reasons for them -- search for 'X'. 

I was wondering about that. The policy seems to be specific to the communication guideline in the Zulip community, e.g. (from https://github.com/zulip/zulip/blob/master/docs/subsystems/emoji.md)
```
* We remove unnecessary gender descriptions. So `:ok_signal:` instead of
  `:ok_woman:`. (strong)
```
But other organizations in Zulip cloud / self-hosted would want to have the variations instead.

Reply:
This PR doesn't add skin tone (which probably requires ZWJ stuff), and so there are not as many variations.

Reply:
> I was wondering about that. The policy seems to be specific to the communication guideline in the Zulip community, e.g. (from https://github.com/zulip/zulip/blob/master/docs/subsystems/emoji.md) … But other organizations in Zulip cloud / self-hosted would want to have the variations instead.

In general the reasoning there is meant as product choices that make Zulip better for its users in general, not as community-specific choices for the chat.zulip.org community. (They aren't uniformly *good* product choices, I think, which is among the motivations for #18121 -- but that's what they're aiming to be.)

For example, this principle:

> We don't add names that could be inappropriate in school or work environments, even if the use is common on the internet. For example, we have not added `:butt:` for `:peach:`, or `:cheers:` for `:beers:`. (strong)

reflects the fact that Zulip is broadly used in school and work environments, and we want to avoid unnecessarily making people uncomfortable about having it there.

> ```
> * We remove unnecessary gender descriptions. So `:ok_signal:` instead of
>   `:ok_woman:`. (strong)
> ```

And this one similarly reflects a product choice: Zulip shouldn't embed gender assumptions, or introduce gender where there's nothing calling for it.

The example here looks to be one that predates combinatorial gender-specifying emoji sequences, where the modern spec explicitly says it's ungendered and if you want a particular gender you say so with a sequence. [Emojipedia says](https://emojipedia.org/person-gesturing-ok/):
> No gender is specified.
> Previously displayed as an alien or blob on Android, a man on Windows, and a woman on most other platforms.
> …
> :ok_woman: (Slack)
> :ok_person: (Github)
> :person_gesturing_ok: (Emojipedia)

And CLDR calls it "person gesturing OK".

I'd call this a case where our existing choice is better than Slack's ("ok_woman", unnecessarily gendered) but not as good as CLDR's, or for that matter GitHub's (`:ok_person:`).

Reply:
> This PR doesn't add skin tone (which probably requires ZWJ stuff), and so there are not as many variations.

Great, that simplifies things.

What in this PR controls that fact? Are those not in emoji-datasource-google in the same way?

Reply:
I didn't specify anything. The json data in iamcal simply doesn't have the skin modifier. Since Matrix.org uses emojibase (which has identical naming as iamcal), they also don't have the skin modifier.